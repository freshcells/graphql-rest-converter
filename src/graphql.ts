import _ from 'lodash'
import {
  DocumentNode,
  GraphQLSchema,
  visit,
  print,
  Kind,
  ASTNode,
  buildASTSchema,
  OperationDefinitionNode,
  GraphQLDirective,
} from 'graphql'
import { gql } from 'graphql-tag'
import { OpenAPIV3 } from 'openapi-types'
import { getVariablesFromPathTemplate } from './pathTemplate'
import {
  createFragmentMap,
  getDependencyClosure,
  getDirective,
  getDirectiveArguments,
  getFragmentDependencies,
  getReferencedFragments,
  isOperationDefinitionNode,
} from './graphqlUtils'
import { CustomProperties, OAType, OpenAPIGraphQLOperation } from './types'
import { getReferenceableFragments, GraphQLTypeToOpenAPITypeSchemaConverter } from './typeConverter'

const DIRECTIVE_DEFINITION = gql`
  input OAExternalDocsInput {
    url: String!
    description: String
  }

  enum HTTPVerb {
    POST
    PUT
    DELETE
  }

  enum ParameterSource {
    PATH
    QUERY
    HEADER
  }

  directive @OAQuery(
    path: String!
    tags: [String!]
    summary: String
    description: String
    externalDocs: OAExternalDocsInput
    deprecated: Boolean
  ) on QUERY

  directive @OAParam(
    in: ParameterSource
    deprecated: Boolean
    description: String
    name: String
  ) on VARIABLE_DEFINITION

  directive @OAMutation(
    path: String!
    tags: [String!]
    summary: String
    description: String
    externalDocs: OAExternalDocsInput
    deprecated: Boolean
    "defaults to POST"
    httpVerb: HTTPVerb
  ) on MUTATION

  directive @OABody(
    "Assume variable name if undefined"
    path: String
    deprecated: Boolean
    description: String
  ) on VARIABLE_DEFINITION

  directive @OAMap(to: String, from: String) on INPUT_FIELD_DEFINITION | VARIABLE_DEFINITION
`

enum OpenAPIDirectives {
  Query = 'OAQuery',
  Param = 'OAParam',
  Mutation = 'OAMutation',
  Body = 'OABody',
  Map = 'OAMap',
}

const graphqlQueryDirectiveDataToOpenAPIOperation = (requestConfig: OpenAPIQueryDirectiveData) => {
  const { tags, summary, description, externalDocs, deprecated } = requestConfig
  return {
    ...(tags?.length ? { tags } : {}),
    ...(typeof summary === 'string' ? { summary } : {}),
    ...(typeof description === 'string' ? { description } : {}),
    ...(typeof externalDocs?.url === 'string' ? { externalDocs } : {}),
    ...(deprecated ? { deprecated } : {}),
  }
}

const getOpenAPIParameters = (
  parameters: Omit<OpenAPIV3.ParameterObject, 'in'>[],
  path: string,
  variablesDirectiveData: Record<string, OpenAPIParamDirectiveData>
) => {
  const adjustedParameters: OpenAPIV3.ParameterObject[] = []
  const variableMap: Record<string, string> = {}

  const pathVariables = new Set(getVariablesFromPathTemplate(path))
  for (const parameter of parameters) {
    const variableName = parameter.name
    const variableDirectiveData = variablesDirectiveData[variableName]
    let parameterName = variableName
    const nameOverride = variableDirectiveData.name
    if (nameOverride) {
      parameterName = nameOverride
    }
    const variableDirectiveDataIn = variableDirectiveData.in?.toLowerCase()

    if (variableDirectiveDataIn) {
      if (pathVariables.has(parameterName) && variableDirectiveDataIn !== 'path') {
        throw new Error(
          `Location ${variableDirectiveDataIn} invalid for parameter ${parameterName} because it is part of the path`
        )
      }
      if (!pathVariables.has(parameterName) && variableDirectiveDataIn === 'path') {
        throw new Error(
          `Location ${variableDirectiveDataIn} invalid for parameter ${parameterName} because it is not part of the path`
        )
      }
      if (variableDirectiveDataIn === 'cookie') {
        throw new Error(`Unsupported parameter location cookie for parameter ${parameterName}`)
      }
      if (!['path', 'query', 'header'].includes(variableDirectiveDataIn)) {
        throw new Error(
          `Unknown parameter location ${variableDirectiveDataIn} for parameter ${parameterName}`
        )
      }
    }
    if (variableDirectiveDataIn && variableDirectiveDataIn === 'header') {
      parameterName = parameterName.toLowerCase()
    }
    if (parameterName !== variableName) {
      variableMap[parameterName] = variableName
    }

    const in_ = pathVariables.has(parameterName)
      ? 'path'
      : variableDirectiveDataIn && ['header', 'cookie'].includes(variableDirectiveDataIn)
      ? variableDirectiveDataIn
      : 'query'
    const deprecated = variableDirectiveData.deprecated
    const description = variableDirectiveData.description
    adjustedParameters.push({
      ...parameter,
      in: in_,
      name: parameterName,
      ...(nameOverride ? { [CustomProperties.VariableName]: parameter.name } : {}),
      ...(deprecated === true ? { deprecated } : {}),
      ...(description ? { description } : {}),
    })
  }

  return {
    parameters: adjustedParameters,
    variableMap,
  }
}

type OpenAPIQueryDirectiveData = {
  path: string
  tags?: [string]
  summary?: string
  description?: string
  externalDocs?: {
    url: string
    description?: string
  }
  deprecated?: boolean
}

type OpenAPIParamDirectiveData = {
  in?: string
  deprecated?: boolean
  description?: string
  name?: string
}

const createOpenAPIOperation = (
  operationId: string | null,
  operationSource: string,
  parameters: OpenAPIV3.ParameterObject[],
  responseSchema: OAType,
  queryDirectiveData: OpenAPIQueryDirectiveData
) => {
  const responses = {
    '200': {
      description: 'Success',
      content: {
        'application/json': {
          schema: responseSchema,
        },
      },
    },
    '400': {
      description: 'Invalid request',
      content: {
        'application/json': {},
      },
    },
  }
  return {
    ...(operationId !== null ? { operationId } : {}),
    [CustomProperties.Operation]: operationSource,
    ...graphqlQueryDirectiveDataToOpenAPIOperation(queryDirectiveData),
    parameters,
    responses,
  }
}

const getVariablesDirectiveData = (
  directive: GraphQLDirective,
  operation: OperationDefinitionNode
) => {
  const result: Record<string, OpenAPIParamDirectiveData> = {}
  for (const variableDefinition of operation.variableDefinitions || []) {
    const paramDirective = getDirective(variableDefinition, OpenAPIDirectives.Param)
    result[variableDefinition.variable.name.value] = paramDirective
      ? (getDirectiveArguments(directive, paramDirective) as OpenAPIParamDirectiveData)
      : {}
  }
  return result
}

export const getOpenAPIGraphQLOperations = (
  schema: GraphQLSchema,
  document: DocumentNode,
  customScalars?: (scalarTypeName: string) => OpenAPIV3.SchemaObject
) => {
  const result: Array<OpenAPIGraphQLOperation> = []

  const fragmentMap = createFragmentMap(document.definitions)

  const fragmentDependencies = getFragmentDependencies(fragmentMap)

  const referencableFragments = getReferenceableFragments(schema, fragmentMap, document)

  const typeConverter = new GraphQLTypeToOpenAPITypeSchemaConverter(
    schema,
    customScalars,
    fragmentMap,
    referencableFragments
  )

  const directiveSchema = buildASTSchema(DIRECTIVE_DEFINITION)

  for (const operation of document.definitions) {
    if (!isOperationDefinitionNode(operation)) {
      continue
    }

    const queryDirective = getDirective(operation, OpenAPIDirectives.Query)
    if (!queryDirective) {
      continue
    }

    const queryDirectiveData = getDirectiveArguments(
      directiveSchema.getDirective(OpenAPIDirectives.Query)!,
      queryDirective
    ) as OpenAPIQueryDirectiveData

    const variablesDirectiveData = getVariablesDirectiveData(
      directiveSchema.getDirective(OpenAPIDirectives.Param)!,
      operation
    )

    const operation_ = removeOpenAPIDirectives(operation)

    const operationFragmentDependencies = getDependencyClosure(
      getReferencedFragments(operation_),
      fragmentDependencies
    )
    const referencedFragments = Object.values(
      _.pickBy(fragmentMap, (v, k) => operationFragmentDependencies.has(k))
    )

    const singleOperationDocument = {
      kind: Kind.DOCUMENT,
      definitions: [operation_, ...referencedFragments],
    } as DocumentNode

    const operationId = operation_.name ? operation_.name.value : null

    const parametersSchema = typeConverter.parametersFromOperation(operation_)

    const responseSchema = typeConverter.responseSchemaFromOperation(operation_)

    const { parameters, variableMap } = getOpenAPIParameters(
      parametersSchema,
      queryDirectiveData.path,
      variablesDirectiveData
    )

    const operationSource = print(singleOperationDocument)

    const openAPIOperation = createOpenAPIOperation(
      operationId,
      operationSource,
      parameters,
      responseSchema,
      queryDirectiveData
    )

    const graphqlOpenAPIOperation = {
      openAPIOperation,
      httpMethod: OpenAPIV3.HttpMethods.GET,
      graphqlDocument: singleOperationDocument,
      variableMap,
      path: queryDirectiveData.path,
    }
    result.push(graphqlOpenAPIOperation)
  }

  return {
    operations: result,
    schemaComponents: typeConverter.getSchemaComponents(),
  }
}

const removeOpenAPIDirectives = <T extends ASTNode>(node: T): T => {
  return visit(node, {
    Directive: {
      enter(node) {
        if (Object.values(OpenAPIDirectives).includes(node.name.value as OpenAPIDirectives)) {
          return null
        }
      },
    },
  })
}
