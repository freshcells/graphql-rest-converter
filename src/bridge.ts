import _ from 'lodash'
import express, { RequestHandler, NextFunction, IRouter } from 'express'
import bodyParser from 'body-parser'
import { OpenAPIV3 } from 'openapi-types'
import { parse, buildSchema, DocumentNode, GraphQLSchema, print, ExecutionResult } from 'graphql'
import OpenAPIRequestCoercer from 'openapi-request-coercer'
import OpenAPIRequestValidator from 'openapi-request-validator'
import OpenAPIResponseValidator from 'openapi-response-validator'
import OpenAPISchemaValidator from 'openapi-schema-validator'
import { PartialDeep } from 'type-fest'
import { getBridgeOperations } from './graphql'
import { pathTemplateToExpressRoute } from './pathTemplate'
import { createOpenAPISchemaFromOperations } from './openApi'
import { CustomProperties, BridgeOperation, BridgeOperations, SchemaComponents } from './types'
import { GraphQLExecutor, GraphQLExecutorArgs } from './graphQLExecutor'

const middlewareToPromise =
  (middleware: RequestHandler) =>
  (req: Parameters<RequestHandler>[0], res: Parameters<RequestHandler>[1]) => {
    return new Promise<ReturnType<RequestHandler>>((resolve, reject) => {
      const next = (x: any) => (x ? reject(x) : resolve())
      middleware(req, res, next)
    })
  }

const promiseToHandler =
  (
    handler: (
      req: Parameters<RequestHandler>[0],
      res: Parameters<RequestHandler>[1]
    ) => Promise<ReturnType<RequestHandler>>
  ) =>
  (...args: Parameters<RequestHandler>) => {
    const result = handler(args[0], args[1])
    result.catch(args[2] as NextFunction)
  }

const jsonBodyParserPromise = middlewareToPromise(bodyParser.json())

const resolveSchemaComponents = (schema: any, schemaComponents: SchemaComponents) => {
  if (typeof schema !== 'object' || schema === null) {
    return
  }
  if ('$ref' in schema) {
    const refTarget = schema.$ref as string
    delete schema.$ref
    for (const [k, v] of Object.entries(
      schemaComponents[refTarget.replace('#/components/schemas/', '')]
    )) {
      schema[k] = _.cloneDeep(v)
    }
  }
  for (const value of Object.values(schema)) {
    resolveSchemaComponents(value, schemaComponents)
  }
}

const addOperation = (
  router: IRouter,
  operation: BridgeOperation,
  schemaComponents: SchemaComponents,
  executor: GraphQLExecutor,
  config: CreateMiddlewareConfig
) => {
  const route = pathTemplateToExpressRoute(operation.path)

  // Resolving `$ref`, at least OpenAPIRequestValidator cannot handle them properly
  const parameters_ = _.cloneDeep(operation.openAPIOperation.parameters || [])
  resolveSchemaComponents(parameters_, schemaComponents)

  const requestCoercer = new OpenAPIRequestCoercer({
    parameters: parameters_,
  })

  const requestValidator =
    typeof config.validateRequest !== 'boolean' || config.validateRequest
      ? new OpenAPIRequestValidator({ parameters: parameters_ })
      : undefined

  const responseValidator = config.validateResponse
    ? new OpenAPIResponseValidator({
        // Type in `openapi-response-validator` seems wrong
        responses: operation.openAPIOperation.responses as any,
        components: { schemas: schemaComponents },
      })
    : undefined

  const graphqlDocument_ = print(operation.graphqlDocument)

  router[operation.httpMethod](
    route,
    promiseToHandler(async (req, res) => {
      if (operation.requestBodyVariable) {
        await jsonBodyParserPromise(req, res)
      }

      const req_ = {
        ...req,
        cookies: { ...req.cookies },
        headers: { ...req.headers },
        params: { ...req.params },
        query: { ...req.query },
        body: req.body,
      }

      // TODO: Cookies not supported
      requestCoercer.coerce(req_)

      const errors = requestValidator ? requestValidator.validateRequest(req_) : null

      if (errors) {
        res.status(400).json(errors)
        return
      }

      const variables: Record<string, any> = {}
      for (const parameter of operation.openAPIOperation
        .parameters as OpenAPIV3.ParameterObject[]) {
        const variableName = operation.variableMap[parameter.name] || parameter.name
        if (parameter.in === 'path') {
          variables[variableName] = req_.params[parameter.name]
        }
        if (parameter.in === 'query') {
          variables[variableName] = req_.query[parameter.name]
        }
        if (parameter.in === 'header') {
          variables[variableName] = req_.headers[parameter.name]
        }
      }
      if (operation.requestBodyVariable) {
        variables[operation.requestBodyVariable] = req_.body
      }

      const request = {
        document: graphqlDocument_,
        variables,
      }
      const result = await executor(request)
      let statusCode = 200
      let contentType: string | undefined = 'application/json'
      let data: string | Buffer | undefined = JSON.stringify(result.data)
      let isTransformed = false

      if (config.responseTransformer) {
        const response_ = config.responseTransformer({
          result,
          request,
          openAPISchema: {
            operation: operation.openAPIOperation,
            method: operation.httpMethod,
            path: operation.path,
          },
        })
        if (response_) {
          statusCode = response_.statusCode
          contentType = response_.contentType
          data = response_.data
          // TODO: The validation can only handle JSON for now
          isTransformed = true
        }
      }

      if (responseValidator && !isTransformed) {
        const responseValidationErrors = responseValidator.validateResponse(statusCode, result.data)
        if (responseValidationErrors) {
          throw new Error(JSON.stringify(responseValidationErrors))
        }
      }

      if (contentType) {
        res.set('Content-Type', contentType)
      }
      res.status(statusCode)
      if (data) {
        res.send(data)
      }
      res.end()
    })
  )
}

export type ResponseTransformerArgs = {
  result: ExecutionResult
  request: GraphQLExecutorArgs
  openAPISchema: {
    operation: OpenAPIV3.OperationObject
    method: OpenAPIV3.HttpMethods
    path: string
  }
}

export type ResponseTransformerResult = {
  statusCode: number
  contentType?: string
  data?: string | Buffer
}

export type ResponseTransformer = (
  args: ResponseTransformerArgs
) => ResponseTransformerResult | undefined

export type CreateMiddlewareConfig = {
  responseTransformer?: ResponseTransformer
  /**
   * Default is `true`
   */
  validateRequest?: boolean
  /**
   * Default is `false`
   */
  validateResponse?: boolean
}

const createExpressMiddleware = (
  operations: BridgeOperations,
  executor: GraphQLExecutor,
  config: CreateMiddlewareConfig
) => {
  // TODO: Avoid depending on express directly?
  const router = express.Router()

  for (const operation of operations.operations) {
    addOperation(router, operation, operations.schemaComponents, executor, config)
  }

  return router
}

type OperationCustomProperties = {
  [CustomProperties.Operation]?: string
}

const getVariableMapFromParameters = (parameters: OpenAPIV3.ParameterObject[]) => {
  const variableMap: Record<string, string> = {}
  for (const parameter of parameters) {
    const variableName = (parameter as any)[CustomProperties.VariableName]
    if (variableName) {
      variableMap[parameter.name] = variableName
    }
  }
  return variableMap
}

const getBridgeOperationsFromOpenAPISchema = (
  schema: OpenAPIV3.Document<OperationCustomProperties>
) => {
  const operations: Array<BridgeOperation> = []
  for (const [path, pathItem] of Object.entries(schema.paths)) {
    if (!pathItem) {
      continue
    }
    for (const [httpMethod, operation] of Object.entries(pathItem)) {
      if (!(operation as any)[CustomProperties.Operation]) {
        continue
      }
      const requestBodyVariable =
        (operation as any)?.requestBody?.[CustomProperties.VariableName] ?? null
      operations.push({
        openAPIOperation: operation as OpenAPIV3.OperationObject,
        path,
        httpMethod: httpMethod as OpenAPIV3.HttpMethods,
        graphqlDocument: parse((operation as any)[CustomProperties.Operation]),
        variableMap: getVariableMapFromParameters((operation as any).parameters || []),
        requestBodyVariable,
      })
    }
  }

  const schemaComponents = (schema?.components?.schemas as SchemaComponents) || {}

  return {
    operations,
    schemaComponents,
  }
}

export const createExpressMiddlewareFromOpenAPISchema = (
  schema: OpenAPIV3.Document<OperationCustomProperties>,
  executor: GraphQLExecutor,
  config: CreateMiddlewareConfig
) => {
  const operations = getBridgeOperationsFromOpenAPISchema(schema)
  return createExpressMiddleware(operations, executor, config)
}

export const removeCustomProperties = (
  schema: OpenAPIV3.Document<OperationCustomProperties>
): OpenAPIV3.Document => {
  const schemaClone = _.cloneDeep(schema)
  for (const pathItem of Object.values(schemaClone.paths)) {
    if (!pathItem) {
      continue
    }
    for (const operation of Object.values(pathItem)) {
      if (typeof operation !== 'object' || operation === null) {
        continue
      }
      delete (operation as any)[CustomProperties.Operation]
      for (const parameter of (operation as any).parameters || []) {
        delete parameter[CustomProperties.VariableName]
      }
    }
  }
  return schemaClone
}

export type CreateOpenAPISchemaConfig = {
  baseSchema: PartialDeep<OpenAPIV3.Document>
  validate?: boolean
}

const createOpenAPISchemaWithValidate = (
  operations: BridgeOperations,
  config: CreateOpenAPISchemaConfig
) => {
  const openAPISchema = createOpenAPISchemaFromOperations(config.baseSchema, operations)
  if (config.validate) {
    const schemaValidator = new OpenAPISchemaValidator({ version: 3 })
    const schemaValidationErrors = schemaValidator.validate(openAPISchema)
    if (schemaValidationErrors?.errors?.length) {
      throw new Error(JSON.stringify(schemaValidationErrors))
    }
  }
  return openAPISchema
}

export type CreateOpenAPIGraphQLBridgeConfig = {
  graphqlSchema: GraphQLSchema | string
  graphqlDocument: DocumentNode | string
  customScalars?: (scalarTypeName: string) => OpenAPIV3.SchemaObject
}

export const createOpenAPIGraphQLBridge = (config: CreateOpenAPIGraphQLBridgeConfig) => {
  const { graphqlSchema, graphqlDocument, customScalars } = config

  const graphqlSchema_ =
    typeof graphqlSchema === 'string' ? buildSchema(graphqlSchema) : graphqlSchema

  const graphqlDocument_ =
    typeof graphqlDocument === 'string' ? parse(graphqlDocument) : graphqlDocument

  const operations = getBridgeOperations(graphqlSchema_, graphqlDocument_, customScalars)

  return {
    getExpressMiddleware: (executor: GraphQLExecutor, config: CreateMiddlewareConfig) =>
      createExpressMiddleware(operations, executor, config),
    getOpenAPISchema: (
      config: CreateOpenAPISchemaConfig
    ): OpenAPIV3.Document<OperationCustomProperties> =>
      createOpenAPISchemaWithValidate(operations, config),
  }
}
