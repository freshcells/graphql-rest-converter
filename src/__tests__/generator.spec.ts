import schema from './schema/schema.graphql'
import { buildASTSchema } from 'graphql'
import { createOpenAPIGraphQLBridge } from '../bridge'
import { gql } from 'graphql-tag'
import { bridgeFixtures } from './bridgeFixtures'
import { getBridgeOperations } from '../graphql'

const graphqlSchema = buildASTSchema(schema)

describe('OpenAPI Generation', () => {
  describe('Bridge schema errors', () => {
    it('should throw for unsupported operations', () => {
      expect(() => {
        getBridgeOperations(
          graphqlSchema,
          gql`
            subscription mySubscription($id: Int!) @OAOperation(path: "/someOp") {
              sampleAdded(id: $id) {
                id
              }
            }
          `
        )
      }).toThrow(/Subscriptions \(at: mySubscription\) are unsupported at this moment/)
    })

    it('should deny multiple @OABody directives', () => {
      expect(() =>
        getBridgeOperations(
          graphqlSchema,
          gql`
            mutation myMutation(
              $inputFirst: SampleInput! @OABody
              $inputSecond: SampleInput! @OABody
            ) @OAOperation(path: "/myMutation") {
              createSampleOne: createSample(input: $inputFirst) {
                id
              }
              createSampleTwo: createSample(input: $inputSecond) {
                id
              }
            }
          `
        )
      ).toThrow(/Only one "OABody" variable allowed/)
    })

    it('should fail in case a `path` parameter is not specified in the path', () => {
      expect(() =>
        getBridgeOperations(
          graphqlSchema,
          gql`
            query myQuery($id: Int! @OAParam(in: PATH)) @OAOperation(path: "/my-query") {
              getSample(id: $id) {
                id
              }
            }
          `
        )
      ).toThrow(/Location path invalid for parameter id because it is not part of the path/)
    })
    it('should fail in case a path parameter is used within another parameter type', () => {
      expect(() =>
        getBridgeOperations(
          graphqlSchema,
          gql`
            query myQuery($id: Int! @OAParam(in: QUERY)) @OAOperation(path: "/my-query/{id}") {
              getSample(id: $id) {
                id
              }
            }
          `
        )
      ).toThrow(/Location query invalid for parameter id because it is part of the path/)
    })
  })

  it('should create schema from graphql api', () => {
    const bridge = createOpenAPIGraphQLBridge({
      graphqlDocument: bridgeFixtures,
      graphqlSchema,
    })

    const schema = bridge.getOpenAPISchema({
      baseSchema: {
        openapi: '3.0.3',
        info: {
          title: 'Sample API',
          version: '1.0.0',
          description: 'My API',
        },
      },
    })
    expect(schema).toMatchSnapshot()
  })
})
