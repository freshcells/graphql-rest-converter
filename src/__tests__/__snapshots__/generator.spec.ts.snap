// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI Generation should create schema from graphql api 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "Sample.MySamples": Object {
        "description": "Description for my MySamples fragment",
        "nullable": true,
        "properties": Object {
          "myId": Object {
            "description": "This is the default \`Sample\` documentation",
            "format": "int32",
            "type": "integer",
          },
          "myName": Object {
            "description": "This is the default \`Sample\` documentation",
            "type": "string",
          },
        },
        "required": Array [
          "myId",
          "myName",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "description": "My API",
    "title": "Sample API",
    "version": "1.0.0",
  },
  "openapi": "3.0.3",
  "paths": Object {
    "/sample": Object {
      "delete": Object {
        "operationId": "removeSample",
        "parameters": Array [
          Object {
            "in": "query",
            "name": "id",
            "required": undefined,
            "schema": Object {
              "format": "int32",
              "type": "integer",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "description": undefined,
                  "properties": Object {
                    "deleteSample": Object {
                      "description": undefined,
                      "nullable": true,
                      "type": "boolean",
                    },
                  },
                  "required": Array [
                    "deleteSample",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "500": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal server error",
          },
        },
        "summary": "Updates a sample",
        "tags": Array [
          "Sample",
        ],
        "x-graphql-operation": "mutation removeSample($id: Int!) {
  deleteSample(id: $id)
}",
      },
      "post": Object {
        "operationId": "createSample",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "description": "This is the default \`SampleInput\` documentation",
                "properties": Object {
                  "moreData": Object {
                    "description": undefined,
                    "items": Object {
                      "description": undefined,
                      "properties": Object {
                        "anotherProperty": Object {
                          "description": undefined,
                          "format": "int32",
                          "nullable": true,
                          "type": "integer",
                        },
                      },
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "name": Object {
                    "description": "The name",
                    "type": "string",
                  },
                },
                "required": Array [
                  "name",
                  "moreData",
                ],
                "type": "object",
              },
            },
          },
          "description": "A Sample Input",
          "x-graphql-variable-name": "sample",
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "description": undefined,
                  "properties": Object {
                    "createSample": Object {
                      "description": "This is the default \`Sample\` documentation",
                      "properties": Object {
                        "id": Object {
                          "description": "This is the default \`Sample\` documentation",
                          "format": "int32",
                          "type": "integer",
                        },
                        "name": Object {
                          "description": "This is the default \`Sample\` documentation",
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "id",
                        "name",
                      ],
                      "type": "object",
                    },
                  },
                  "required": Array [
                    "createSample",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "500": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal server error",
          },
        },
        "summary": "Creates a new sample",
        "tags": Array [
          "Sample",
        ],
        "x-graphql-operation": "mutation createSample($sample: SampleInput!) {
  createSample(input: $sample) {
    id
    name
  }
}",
      },
    },
    "/sample/{id}": Object {
      "get": Object {
        "description": "Get a specific sample",
        "operationId": "getSample",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "format": "int32",
              "type": "integer",
            },
          },
        ],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "description": undefined,
                  "properties": Object {
                    "getSample": Object {
                      "allOf": Array [
                        Object {
                          "$ref": "#/components/schemas/Sample.MySamples",
                        },
                      ],
                      "description": "This is the default \`Sample\` documentation",
                      "type": "object",
                    },
                  },
                  "required": Array [
                    "getSample",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "500": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal server error",
          },
        },
        "x-graphql-operation": "query getSample($id: Int!) {
  getSample(id: $id) {
    ...MySamples
  }
}

fragment MySamples on Sample @OADescription(description: \\"Description for my MySamples fragment\\") {
  myId: id @OADescription(description: \\"my custom description on a field\\")
  myName: name
}",
      },
      "put": Object {
        "operationId": "updateSample",
        "parameters": Array [
          Object {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": Object {
              "format": "int32",
              "type": "integer",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "description": "This is the default \`SampleInput\` documentation",
                "properties": Object {
                  "moreData": Object {
                    "description": undefined,
                    "items": Object {
                      "description": undefined,
                      "properties": Object {
                        "anotherProperty": Object {
                          "description": undefined,
                          "format": "int32",
                          "nullable": true,
                          "type": "integer",
                        },
                      },
                      "type": "object",
                    },
                    "type": "array",
                  },
                  "name": Object {
                    "description": "The name",
                    "type": "string",
                  },
                },
                "required": Array [
                  "name",
                  "moreData",
                ],
                "type": "object",
              },
            },
          },
          "x-graphql-variable-name": "sample",
        },
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "description": undefined,
                  "properties": Object {
                    "updateSample": Object {
                      "description": "This is the default \`Sample\` documentation",
                      "properties": Object {
                        "name": Object {
                          "description": "This is the default \`Sample\` documentation",
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "name",
                      ],
                      "type": "object",
                    },
                  },
                  "required": Array [
                    "updateSample",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "500": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal server error",
          },
        },
        "summary": "Updates a sample",
        "tags": Array [
          "Sample",
        ],
        "x-graphql-operation": "mutation updateSample($id: Int!, $sample: SampleInput!) {
  updateSample(id: $id, input: $sample) {
    name
  }
}",
      },
    },
    "/samples": Object {
      "get": Object {
        "description": "Get all samples",
        "operationId": "getSamples",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "description": undefined,
                  "properties": Object {
                    "samples": Object {
                      "description": undefined,
                      "items": Object {
                        "allOf": Array [
                          Object {
                            "$ref": "#/components/schemas/Sample.MySamples",
                          },
                        ],
                        "description": "This is the default \`Sample\` documentation",
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "samples",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
          "400": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Invalid request",
          },
          "500": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "properties": Object {
                    "errors": Object {
                      "items": Object {
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "status": Object {
                      "format": "int32",
                      "type": "number",
                    },
                  },
                  "type": "object",
                },
              },
            },
            "description": "Internal server error",
          },
        },
        "x-graphql-operation": "query getSamples {
  samples: getSamples {
    ...MySamples
  }
}

fragment MySamples on Sample @OADescription(description: \\"Description for my MySamples fragment\\") {
  myId: id @OADescription(description: \\"my custom description on a field\\")
  myName: name
}",
      },
    },
  },
}
`;
